font-face must accept ranges in reverse order, and reverse them for *computed* style

darinadler opened the pull request:
#### 30bf0f961d694189c96d527c0a0c86fe3a656c5d
<pre>
font-face must accept ranges in reverse order, and reverse them for *computed* style
<a href="https://bugs.webkit.org/show_bug.cgi?id=246585">https://bugs.webkit.org/show_bug.cgi?id=246585</a>
rdar://problem/101210777

[ Mac ] imported/w3c/web-platform-tests/server-timing/service_worker_idl.html is flaky failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=219907">https://bugs.webkit.org/show_bug.cgi?id=219907</a>
rdar://problem/72351272

<a href="https://bugs.webkit.org/show_bug.cgi?id=244900">https://bugs.webkit.org/show_bug.cgi?id=244900</a>
New Test(254036@main): [ iOS macOS Debug ] 3X imported/w3c/web-platform-tests/dom/nodes/NodeList-static-length-getter-tampered-indexOf-(Layout tests) are constant timeouts
rdar://problem/99657326

Reviewed by Tim Nguyen.

The biggest change here is to store the specified values in CSSFontFace. The existing code was
not storing these values, and when querying it would return a computed value, but that is not
the same thing. We store the values either in the style properties of a StyleRuleFontFace if this
font face is connected to a CSS style rule, or in our own MutableStyleProperties object if it is
not connected.

When it comes to tests, many of the ones we wrote for WebKit were expecting the computed values,
but the web platform tests were expecting the specified values. The WebKit ones needed some
updating to expect the correct behavior.

* LayoutTests/fast/text/font-face-javascript-expected.txt:
* LayoutTests/fast/text/font-face-javascript.html: Updated to expect the style attribute to return
the specified style &quot;oblique&quot; rather than the computed style &quot;italic&quot;.

* LayoutTests/fast/text/font-loading-font-display-expected.txt:
* LayoutTests/fast/text/font-loading-font-display.html: Updated to expect font-display to get set
to &quot;auto&quot; when we set the display attribute on a FontFace to &quot;auto&quot;. The old code expected the
empty string, but a review of the specification makes it clear that was not correct.

* LayoutTests/fast/text/font-selection-font-face-parse-expected.txt:
* LayoutTests/fast/text/font-selection-font-face-parse.html: Updated since out of order ranges now
are parsed successfully. There was one test case each for fontWeight, fontStretch, and fontStyle
that was affected.

* LayoutTests/fast/text/font-selection-font-loading-api-parse-expected.txt:
* LayoutTests/fast/text/font-selection-font-loading-api-parse.html: Updated since the style
attribute now returns the specified style rather than the computed style. That means we get &quot;400&quot;
instead of &quot;normal&quot; and &quot;700&quot; instead of &quot;bold&quot;, and other similar changes.

* LayoutTests/fast/text/variations/font-loading-api-parse-ranges-expected.txt:
* LayoutTests/fast/text/variations/font-loading-api-parse-ranges.html: Updated since out of order
ranges are now parsed successfully. There was one case each for weight, stretch, and style that
was affected.

* LayoutTests/TestExpectations: Removed expectation for simulcast-h264.html, which has been removed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-face-range-order-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/at-font-face-descriptors-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/at-font-face-font-matching-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/font-parse-numeric-stretch-style-weight-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/font-style-parsing-expected.txt:
Updated expected results. These are mostly progressions. In a few cases we have a test that is
newly failing because we are now parsing font-face correctly and matching incorrectly. In those
cases the old version failed to parse so the test was saying PASS but that was a false negative.

* LayoutTests/platform/gtk/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.
* LayoutTests/platform/ios/fast/text/font-selection-font-face-parse-expected.txt: Removed.
* LayoutTests/platform/ios/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.
* LayoutTests/platform/mac/fast/text/font-selection-font-face-parse-expected.txt: Removed.
* LayoutTests/platform/mac/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.
* LayoutTests/platform/wpe/fast/text/font-selection-font-face-parse-expected.txt: Removed.
* LayoutTests/platform/wpe/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.

* LayoutTests/platform/mac/TestExpectations: Removed expectations for service_worker_idl.html and
NodeList-static-length-getter-tampered-indexOf-[1-3].html, which have been removed.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::propertiesOrCSSConnection): Added. Helper for the below. Could have used a lambda, but then
the style checker would have gotten all confused.
(WebCore::CSSFontFace::CSSFontFace): Initialize m_propertiesOrCSSConnection.
(WebCore::CSSFontFace::properties): Added.
(WebCore::CSSFontFace::mutableProperties): Added.
(WebCore::CSSFontFace::cssConnection): Moved here from the header.
(WebCore::CSSFontFace::setFamilies): Changed argument type to CSSValueList. Use mutableProperties.
Removed unneeded boolean return value.
(WebCore::CSSFontFace::setWeight): Use mutableProperties. Set the weight in
m_fontSelectionCapabilities directly rather than using a helper function.
(WebCore::CSSFontFace::setStretch): Use mutableProperties. Set the stretch in
m_fontSelectionCapabilities directly rather than using a helper function.
(WebCore::CSSFontFace::setStyle): Use mutableProperties. Set the slope in
m_fontSelectionCapabilities directly rather than using a helper function.
(WebCore::CSSFontFace::setUnicodeRange): Changed argument type to CSSValueList. Use
mutableProperties. Removed unneeded boolean return value.
(WebCore::CSSFontFace::setFeatureSettings): Use mutableProperties. Renamed the
existing data member to m_parsedFeatureSettings.
(WebCore::CSSFontFace::setDisplay): Renamed from setLoadingBehavior to match the terminology in the
specification. Use mutableProperties. Changed the argument type to CSSPrimitiveValue.
(WebCore::CSSFontFace::family const): Added. Gets the property from the properties object. Moved
logic here, to work around the fact that we use a list, from FontFace::family.
(WebCore::CSSFontFace::style const): Ditto.
(WebCore::CSSFontFace::weight const): Ditto.
(WebCore::CSSFontFace::stretch const): Ditto.
(WebCore::CSSFontFace::unicodeRange const): Ditto.
(WebCore::CSSFontFace::featureSettings const): Ditto.
(WebCore::CSSFontFace::display const): Ditto.
(WebCore::CSSFontFace::families const): Just a simple getter, here so the header does not have
to include CSSValueList.h.

* Source/WebCore/css/CSSFontFace.h: Updated for the above. Removed the functions that returned
various computed properties, and instead have functions that return the specified properties
as strings since that is what FontFace needs for its interface. The only ones that return computed
properties that are kept are families, ranges, and fontSelectionCapabilities; they could probably
use names that make it clear they return computed values. Return a span instead of a vector from
the ranges function, a pointer instead of an optional pointer from the families function, and a
non-optional FontSelectionCapabilities from the fontSelectionCapabilities function. Added private
member functions named properties and mutableProperties and replaced m_cssConnection with a
private data member named m_propertiesOrCSSConnection. These store the specified property values
in the appropriate place whether or not a CSS connection exists.

* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::ensureLocalFontFacesForFamilyRegistered): Added a comment.
(WebCore::CSSFontFaceSet::addToFacesLookupTable): Updated for the return type change of
CSSFontFace::families.
(WebCore::CSSFontFaceSet::remove): Ditto.
(WebCore::CSSFontFaceSet::fontFace): Updated to check the loading staus explicitly instead of
doing it implicitly with the optional status of fontSelectionCapabilities. This makes it more
explicit what this is doing, and perhaps might even lead us to discover the code is unneeded later.
Updated since fontSelectionCapabilities return type is no longer optional.

* Source/WebCore/css/CSSFontFaceSet.h: Added newly-needed forward declaration.

* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontFaceRule): Updated for changes to CSSFontFace::setFamilies,
CSSFontFace::setUnicodeRange, and CSSFontFace::setDisplay.

* Source/WebCore/css/CSSFontSelector.h: Removed unneeded forward declaration.

* Source/WebCore/css/CSSSegmentedFontFace.cpp:
(WebCore::CSSSegmentedFontFace::fontRanges): Updated for changes to CSSFontFace::ranges and
CSSFontFace::fontSelectionCapabilities.

* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::setFamily): Updated for change to CSSFontFace::setFamilies.
(WebCore::FontFace::setStyle): Removed unneeded special case for empty string; the
CSSPropertyParserWorkerSafe::parseFontFaceStyle function fails to parse it.
(WebCore::FontFace::setWeight): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceWeight function fails to parse it.
(WebCore::FontFace::setStretch): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceStretch function fails to parse it.
(WebCore::FontFace::setUnicodeRange): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceUnicodeRange function fails to parse it. Also updated
for change to CSSFontFace::setUnicodeRange.
(WebCore::FontFace::setFeatureSettings): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceFeatureSettings function fails to parse it. Changed
the coding style to match the other surrounding functions.
(WebCore::FontFace::setDisplay): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceDisplay function fails to parse it. Updated for the
change to CSSFontFace::setDisplay.
(WebCore::FontFace::family const): Instead of a computed value string, get the specified value
string by calling the function on the CSSFontFace.
(WebCore::FontFace::style const): Ditto.
(WebCore::FontFace::weight const): Ditto.
(WebCore::FontFace::stretch const): Ditto.
(WebCore::FontFace::unicodeRange const): Ditto.
(WebCore::FontFace::featureSettings const): Ditto.
(WebCore::FontFace::display const): Ditto.

* Source/WebCore/css/FontFace.h: Removed ScriptExecutionContext argument from display getter.
* Source/WebCore/css/FontFace.idl: Updated for the above.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
(WebCore::CSSPropertyParserHelpers::isFontStyleAngleInRange): Updated to allow -90 and 90
degrees as specified.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserWorkerSafe::parseFontFaceUnicodeRange): Return CSSValueList.
(WebCore::CSSPropertyParserWorkerSafe::parseFontFaceDisplay): Return CSSPrimitiveValue.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontStyleRange): Removed the check that the
two angles are in order. This is now handled as part of style computation instead.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontWeightAbsoluteRange): Removed the check
that the two weights are in order. This is now handled as part of style computation instead.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontStretchRange): Removed the check that the
two stretches are in order. This is now handled as part of style computation instead.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h: Updated the return types mentioned above.

* Source/WebCore/platform/graphics/FontSelectionAlgorithm.h: Change the FontSelectionRange
constructor to sort the two values so either can be the minimum and either can be the maximum.
This is the behavior the specification calls for. And it&apos;s required we do this sorting as part
of the computation process, rather than when parsing the specified values.

Canonical link: <a href="https://commits.webkit.org/255893@main">https://commits.webkit.org/255893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7256d6a326935ef858b0a5c713ee0bbf174ce192

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24486 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103548 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163886 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97908 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3123 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31343 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86238 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99568 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99581 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2233 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80339 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29251 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84159 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72205 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37736 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17688 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35602 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18951 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4068 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39479 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41519 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41415 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38181 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSFontFace.h:
```diff
@@ -84,22 +87,18 @@ class CSSFontFace final : public RefCounted<CSSFontFace> {
+    FontSelectionCapabilities fontSelectionCapabilities() const { ASSERT(m_status != Status::Failure); return m_fontSelectionCapabilities.computeFontSelectionCapabilities(); }
+
+    // FIXME: Consider adding the word "specified" to this function's name.
     void setFontSelectionCapabilities(FontSelectionCapabilities capabilities) { m_fontSelectionCapabilities = capabilities; }
```
Could you file a bug for these FIXME?

darinadler replied:
I’d rather take them out. I don’t think this ideas are so good that they should have bug reports.

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
```diff
@@ -247,7 +247,7 @@ template<CSSValueID... names> RefPtr<CSSPrimitiveValue> consumeIdentWorkerSafe(C
 inline bool isFontStyleAngleInRange(double angleInDegrees)
 {
-    return angleInDegrees > -90 && angleInDegrees < 90;
+    return angleInDegrees >= -90 && angleInDegrees <= 90;
```
You might want to update the changelog now that I made the change in a different commit.

darinadler replied:
OK, will do.

nt1m commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -95,8 +95,16 @@ Ref<CSSFontFace> CSSFontFace::create(CSSFontSelector& fontSelector, StyleRuleFon
 CSSFontFace::CSSFontFace(const Settings::Values* settings, StyleRuleFontFace* cssConnection, FontFace* wrapper, bool isLocalFallback)
-    : m_cssConnection(cssConnection)
+    : m_propertiesUnlessConnected(initialStyleProperties(cssConnection))
+    , m_cssConnection(cssConnection)
```
The name `m_propertiesUnlessConnected` is a bit confusing, for someone reading this class at first sight, they might think a connection can be added to a `CSSFontFace` after it's been initialized, which is not the case. 

I'm not sure what would be a good name though, maybe `m_propertiesIfNotConnected` ?

darinadler replied:
I’d be happy to change the name.

The existing `m_cssConnection` actually did make me think a connection could be added to `CSSFontFace` after it’s been initialized, so that concern is not new. I could use `std::variant` to combine `m_propertiesUnlessConnected` and `m_cssConnection` and make that data member `const`, but then what would I name it.

nt1m replied:
I think it would help if `initialStyleProperties` was inlined: `m_propertiesIfNotConnected(cssConnection ? nullptr : MutableStyleProperties::create())` 

The combination of `initial` and `unless` make it confusing I think.

darinadler replied:
I have a new version that uses `std::variant`. Came out pretty good.

nt1m commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -95,8 +95,16 @@ Ref<CSSFontFace> CSSFontFace::create(CSSFontSelector& fontSelector, StyleRuleFon
+
+CSSFontFace::CSSFontFace(const Settings::Values* settings, StyleRuleFontFace* connection, FontFace* wrapper, bool isLocalFallback)
+    : m_propertiesOrCSSConnection(propertiesOrCSSConnection(connection))
+    , m_cssConnection(cssConnection)
```
I guess you meant to remove `m_cssConnection(cssConnection)`

darinadler committed:
font-face must accept ranges in reverse order, and reverse them for *computed* style
https://bugs.webkit.org/show_bug.cgi?id=246585
rdar://problem/101210777

[ Mac ] imported/w3c/web-platform-tests/server-timing/service_worker_idl.html is flaky failing
https://bugs.webkit.org/show_bug.cgi?id=219907
rdar://problem/72351272

https://bugs.webkit.org/show_bug.cgi?id=244900
New Test(254036@main): [ iOS macOS Debug ] 3X imported/w3c/web-platform-tests/dom/nodes/NodeList-static-length-getter-tampered-indexOf-(Layout tests) are constant timeouts
rdar://problem/99657326

Reviewed by Tim Nguyen.

The biggest change here is to store the specified values in CSSFontFace. The existing code was
not storing these values, and when querying it would return a computed value, but that is not
the same thing. We store the values either in the style properties of a StyleRuleFontFace if this
font face is connected to a CSS style rule, or in our own MutableStyleProperties object if it is
not connected.

When it comes to tests, many of the ones we wrote for WebKit were expecting the computed values,
but the web platform tests were expecting the specified values. The WebKit ones needed some
updating to expect the correct behavior.

* LayoutTests/fast/text/font-face-javascript-expected.txt:
* LayoutTests/fast/text/font-face-javascript.html: Updated to expect the style attribute to return
the specified style "oblique" rather than the computed style "italic".

* LayoutTests/fast/text/font-loading-font-display-expected.txt:
* LayoutTests/fast/text/font-loading-font-display.html: Updated to expect font-display to get set
to "auto" when we set the display attribute on a FontFace to "auto". The old code expected the
empty string, but a review of the specification makes it clear that was not correct.

* LayoutTests/fast/text/font-selection-font-face-parse-expected.txt:
* LayoutTests/fast/text/font-selection-font-face-parse.html: Updated since out of order ranges now
are parsed successfully. There was one test case each for fontWeight, fontStretch, and fontStyle
that was affected.

* LayoutTests/fast/text/font-selection-font-loading-api-parse-expected.txt:
* LayoutTests/fast/text/font-selection-font-loading-api-parse.html: Updated since the style
attribute now returns the specified style rather than the computed style. That means we get "400"
instead of "normal" and "700" instead of "bold", and other similar changes.

* LayoutTests/fast/text/variations/font-loading-api-parse-ranges-expected.txt:
* LayoutTests/fast/text/variations/font-loading-api-parse-ranges.html: Updated since out of order
ranges are now parsed successfully. There was one case each for weight, stretch, and style that
was affected.

* LayoutTests/TestExpectations: Removed expectation for simulcast-h264.html, which has been removed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/font-face-range-order-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/at-font-face-descriptors-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/at-font-face-font-matching-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/font-parse-numeric-stretch-style-weight-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/variations/font-style-parsing-expected.txt:
Updated expected results. These are mostly progressions. In a few cases we have a test that is
newly failing because we are now parsing font-face correctly and matching incorrectly. In those
cases the old version failed to parse so the test was saying PASS but that was a false negative.

* LayoutTests/platform/gtk/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.
* LayoutTests/platform/ios/fast/text/font-selection-font-face-parse-expected.txt: Removed.
* LayoutTests/platform/ios/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.
* LayoutTests/platform/mac/fast/text/font-selection-font-face-parse-expected.txt: Removed.
* LayoutTests/platform/mac/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.
* LayoutTests/platform/wpe/fast/text/font-selection-font-face-parse-expected.txt: Removed.
* LayoutTests/platform/wpe/fast/text/font-selection-font-loading-api-parse-expected.txt: Removed.

* LayoutTests/platform/mac/TestExpectations: Removed expectations for service_worker_idl.html and
NodeList-static-length-getter-tampered-indexOf-[1-3].html, which have been removed.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::propertiesOrCSSConnection): Added. Helper for the below. Could have used a lambda, but then
the style checker would have gotten all confused.
(WebCore::CSSFontFace::CSSFontFace): Initialize m_propertiesOrCSSConnection.
(WebCore::CSSFontFace::properties): Added.
(WebCore::CSSFontFace::mutableProperties): Added.
(WebCore::CSSFontFace::cssConnection): Moved here from the header.
(WebCore::CSSFontFace::setFamilies): Changed argument type to CSSValueList. Use mutableProperties.
Removed unneeded boolean return value.
(WebCore::CSSFontFace::setWeight): Use mutableProperties. Set the weight in
m_fontSelectionCapabilities directly rather than using a helper function.
(WebCore::CSSFontFace::setStretch): Use mutableProperties. Set the stretch in
m_fontSelectionCapabilities directly rather than using a helper function.
(WebCore::CSSFontFace::setStyle): Use mutableProperties. Set the slope in
m_fontSelectionCapabilities directly rather than using a helper function.
(WebCore::CSSFontFace::setUnicodeRange): Changed argument type to CSSValueList. Use
mutableProperties. Removed unneeded boolean return value.
(WebCore::CSSFontFace::setFeatureSettings): Use mutableProperties. Renamed the
existing data member to m_parsedFeatureSettings.
(WebCore::CSSFontFace::setDisplay): Renamed from setLoadingBehavior to match the terminology in the
specification. Use mutableProperties. Changed the argument type to CSSPrimitiveValue.
(WebCore::CSSFontFace::family const): Added. Gets the property from the properties object. Moved
logic here, to work around the fact that we use a list, from FontFace::family.
(WebCore::CSSFontFace::style const): Ditto.
(WebCore::CSSFontFace::weight const): Ditto.
(WebCore::CSSFontFace::stretch const): Ditto.
(WebCore::CSSFontFace::unicodeRange const): Ditto.
(WebCore::CSSFontFace::featureSettings const): Ditto.
(WebCore::CSSFontFace::display const): Ditto.
(WebCore::CSSFontFace::families const): Just a simple getter, here so the header does not have
to include CSSValueList.h.

* Source/WebCore/css/CSSFontFace.h: Updated for the above. Removed the functions that returned
various computed properties, and instead have functions that return the specified properties
as strings since that is what FontFace needs for its interface. The only ones that return computed
properties that are kept are families, ranges, and fontSelectionCapabilities; they could probably
use names that make it clear they return computed values. Return a span instead of a vector from
the ranges function, a pointer instead of an optional pointer from the families function, and a
non-optional FontSelectionCapabilities from the fontSelectionCapabilities function. Added private
member functions named properties and mutableProperties and replaced m_cssConnection with a
private data member named m_propertiesOrCSSConnection. These store the specified property values
in the appropriate place whether or not a CSS connection exists.

* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::ensureLocalFontFacesForFamilyRegistered): Added a comment.
(WebCore::CSSFontFaceSet::addToFacesLookupTable): Updated for the return type change of
CSSFontFace::families.
(WebCore::CSSFontFaceSet::remove): Ditto.
(WebCore::CSSFontFaceSet::fontFace): Updated to check the loading staus explicitly instead of
doing it implicitly with the optional status of fontSelectionCapabilities. This makes it more
explicit what this is doing, and perhaps might even lead us to discover the code is unneeded later.
Updated since fontSelectionCapabilities return type is no longer optional.

* Source/WebCore/css/CSSFontFaceSet.h: Added newly-needed forward declaration.

* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::addFontFaceRule): Updated for changes to CSSFontFace::setFamilies,
CSSFontFace::setUnicodeRange, and CSSFontFace::setDisplay.

* Source/WebCore/css/CSSFontSelector.h: Removed unneeded forward declaration.

* Source/WebCore/css/CSSSegmentedFontFace.cpp:
(WebCore::CSSSegmentedFontFace::fontRanges): Updated for changes to CSSFontFace::ranges and
CSSFontFace::fontSelectionCapabilities.

* Source/WebCore/css/FontFace.cpp:
(WebCore::FontFace::setFamily): Updated for change to CSSFontFace::setFamilies.
(WebCore::FontFace::setStyle): Removed unneeded special case for empty string; the
CSSPropertyParserWorkerSafe::parseFontFaceStyle function fails to parse it.
(WebCore::FontFace::setWeight): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceWeight function fails to parse it.
(WebCore::FontFace::setStretch): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceStretch function fails to parse it.
(WebCore::FontFace::setUnicodeRange): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceUnicodeRange function fails to parse it. Also updated
for change to CSSFontFace::setUnicodeRange.
(WebCore::FontFace::setFeatureSettings): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceFeatureSettings function fails to parse it. Changed
the coding style to match the other surrounding functions.
(WebCore::FontFace::setDisplay): Removed unneeded special case for empty string, since the
CSSPropertyParserWorkerSafe::parseFontFaceDisplay function fails to parse it. Updated for the
change to CSSFontFace::setDisplay.
(WebCore::FontFace::family const): Instead of a computed value string, get the specified value
string by calling the function on the CSSFontFace.
(WebCore::FontFace::style const): Ditto.
(WebCore::FontFace::weight const): Ditto.
(WebCore::FontFace::stretch const): Ditto.
(WebCore::FontFace::unicodeRange const): Ditto.
(WebCore::FontFace::featureSettings const): Ditto.
(WebCore::FontFace::display const): Ditto.

* Source/WebCore/css/FontFace.h: Removed ScriptExecutionContext argument from display getter.
* Source/WebCore/css/FontFace.idl: Updated for the above.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
(WebCore::CSSPropertyParserHelpers::isFontStyleAngleInRange): Updated to allow -90 and 90
degrees as specified.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserWorkerSafe::parseFontFaceUnicodeRange): Return CSSValueList.
(WebCore::CSSPropertyParserWorkerSafe::parseFontFaceDisplay): Return CSSPrimitiveValue.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontStyleRange): Removed the check that the
two angles are in order. This is now handled as part of style computation instead.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontWeightAbsoluteRange): Removed the check
that the two weights are in order. This is now handled as part of style computation instead.
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontStretchRange): Removed the check that the
two stretches are in order. This is now handled as part of style computation instead.

* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.h: Updated the return types mentioned above.

* Source/WebCore/platform/graphics/FontSelectionAlgorithm.h: Change the FontSelectionRange
constructor to sort the two values so either can be the minimum and either can be the maximum.
This is the behavior the specification calls for. And it's required we do this sorting as part
of the computation process, rather than when parsing the specified values.

Canonical link: https://commits.webkit.org/255893@main
